backends-velox/src/test/scala/org/apache/gluten/execution/TestOperator.scala

```diff
@@ -1236,4 +1236,34 @@ class TestOperator extends VeloxWholeStageTransformerSuite {
       }
     }
   }
+
+  test("Cast date to string") {
+    withTempPath {
+      path =>
+        Seq("2023-01-01", "2023-01-02", "2023-01-03")
+          .toDF("dateColumn")
+          .select(to_date($"dateColumn", "yyyy-MM-dd").as("dateColumn"))
+          .write
+          .parquet(path.getCanonicalPath)
+        spark.read.parquet(path.getCanonicalPath).createOrReplaceTempView("view")
+        runQueryAndCompare("SELECT cast(dateColumn as string) from view") {
+          checkGlutenOperatorMatch[ProjectExecTransformer]
+        }
+    }
+  }
+
+  test("Cast date to timestamp") {
```

Comment:
Based on the implementation of Velox, could we verify if the behaviors are aligned with Spark on below two points?

1. With timezone.
2. Velox converts days as millis, then as timestamp. What unit does Spark use in the conversion, milli or micro, could we confirm?

https://github.com/facebookincubator/velox/blob/main/velox/expression/CastExpr.cpp#L181-L198

Reply:
>

Reply:
> With timezone 

I believe it's ok. Spark's conversion of the date type to timestamp is only supported up to the day of the day, and will not cause problems in different time zones.  
```
scala> val df = spark.sql("select date'2023-01-02 01:01:01' as ts")
df: org.apache.spark.sql.DataFrame = [ts: date]

scala> df.show
+----------+
|        ts|
+----------+
|2023-01-02|
+----------+

scala> spark.sql("set spark.sql.session.timeZone = America/Los_Angeles")
res22: org.apache.spark.sql.DataFrame = [key: string, value: string]

scala> spark.sql("select cast(date'2023-01-02 01:01:01' as timestamp) as ts").show
+-------------------+
|                 ts|
+-------------------+
|2023-01-02 00:00:00|
+-------------------+


scala> spark.sql("set spark.sql.session.timeZone = Asia/Shanghai")
res24: org.apache.spark.sql.DataFrame = [key: string, value: string]

scala> spark.sql("select cast(date'2023-01-02 01:01:01' as timestamp) as ts").show
+-------------------+
|                 ts|
+-------------------+
|2023-01-02 00:00:00|
+-------------------+

```

Reply:
Thanks for checking. Do we need to add some tests with timezones to make sure the behaviors are aligned?

Reply:
> Spark's conversion of the date type to timestamp is only supported up to the day of the day, and will not cause problems in different time zones.
spark.sql("select cast(date'2023-01-02 01:01:01' as timestamp) as ts").show
+-------------------+
|                 ts          |
+-------------------+
|2023-01-02 00:00:00|
+-------------------+

@dcoliversun, let me help clarify a bit. Actually, timezone matters in casting date to timestamp, i.e., date value is adjusted according to the configured local timezone during the casting. And in Spark, timestamp is always corresponding to UTC+0 timezone.
The reason for why the above sql returns the original input value is, the final result is produced by implicitly casting timestamp to string, where local timezone is also considered. See [spark code](https://github.com/apache/spark/blob/v3.3.4/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L358). This can also explain your mentioned phenomenon: "In spark, timestamp behaves differently in df.show and df.collect".

We can let Spark write timestamp result to parquet, and then print the parquet content to see the difference when different timezone is configured. Or, just check the difference of returned timestamp dataframe, as your added test does.